### PR TITLE
Add debug expression evaluation support for equality and type check expressions

### LINKED ISSUE #26321: [Debugger] Add expression evaluation support for equality and type check expressions
**Description:**
The existing evaluation engine implementation should be improved to support the following expression types.
- [x] Equality Expressions
  - [x] Deep Value Equality
  - [x] Reference Equality
- [x] Type test Expressions

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by NipunaRanasinghe**
## Purpose

This PR adds ballerina debug expression evaluation support for
- [x] Equality Expressions
  - [x] Deep Value Equality
  - [x] Reference Equality
- [x] Type test Expressions

This PR also adds integration test scenarios for the added expression types.

Fixes #26321.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [x] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/EvaluationUtils.java [hunk 1]
```diff
@@ -36,13 +41,57 @@
...
+    public static final String JAVA_BOOLEAN_CLASS = "java.lang.Boolean";
+    public static final String JAVA_LONG_CLASS = "java.lang.Long";
+    public static final String JAVA_DOUBLE_CLASS = "java.lang.Double";
+    public static final String JAVA_LANG_CLASS = "java.lang.Class";
+    public static final String JAVA_OBJECT_CLASS = "java.lang.Object";
+    // Helper methods
+    public static final String GET_TYPEDESC_METHOD = "getTypedesc";
+    public static final String VALUE_OF_METHOD = "valueOf";
+    public static final String REF_EQUAL_METHOD = "isReferenceEqual";
+    public static final String VALUE_EQUAL_METHOD = "isEqual";
```
Comment (Ibaqu) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/EvaluationUtils.java:56]: Can these be private? These are used only within the EvaluationUtils class right?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java [hunk 2]
```diff
@@ -104,6 +114,12 @@ private BExpressionValue performOperation(BExpressionValue lhs, BExpressionValue
                 return logicalOR(lVar, rVar);
             case ELVIS_TOKEN:
                 return conditionalReturn(lVar, rVar);
+            case DOUBLE_EQUAL_TOKEN:
+            case NOT_EQUAL_TOKEN:
+                return checkValueEquality(lVar, rVar, operatorType);
+            case TRIPPLE_EQUAL_TOKEN:
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java:120]: What is the relevant action if the operatorType is `DOUBLE_EQUAL_TOKEN` and TRIPPLE_EQUAL_TOKEN?
Are we planning to implement these in future?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java [hunk 3]
```diff
@@ -387,6 +403,50 @@ private BExpressionValue conditionalReturn(BVariable lVar, BVariable rVar) {
...
+     */
+    private BExpressionValue checkValueEquality(BVariable lVar, BVariable rVar, SyntaxKind operatorType)
+            throws EvaluationException {
+        List<Value> argList = new ArrayList<>();
+        argList.add(getValueAsObject(context, lVar));
+        argList.add(getValueAsObject(context, rVar));
+
+        List<String> argTypeNames = new ArrayList<>();
+        argTypeNames.add(JAVA_OBJECT_CLASS);
+        argTypeNames.add(JAVA_OBJECT_CLASS);
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java:417]: Seems like you are adding the `JAVA_OBJECT_CLASS` twice. This is redundant.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java [hunk 4]
```diff
@@ -387,6 +403,50 @@ private BExpressionValue conditionalReturn(BVariable lVar, BVariable rVar) {
...
+     */
+    private BExpressionValue checkReferenceEquality(BVariable lVar, BVariable rVar, SyntaxKind operatorType)
+            throws EvaluationException {
+        List<Value> argList = new ArrayList<>();
+        argList.add(getValueAsObject(context, lVar));
+        argList.add(getValueAsObject(context, rVar));
+
+        List<String> argTypeNames = new ArrayList<>();
+        argTypeNames.add(JAVA_OBJECT_CLASS);
+        argTypeNames.add(JAVA_OBJECT_CLASS);
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java:439]: Same [here](https://github.com/ballerina-platform/ballerina-lang/pull/26322/files#r505441120) as well. Is there any specific reason for this?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmInstanceMethod.java [hunk 5]
```diff
@@ -57,8 +65,44 @@ public Value invoke() throws EvaluationException {
...
+
+            // Todo - IMPORTANT: Add remaining steps to validate and match named, defaultable and rest args
+            // Todo - verify
+            // Here we use the parent strand instance to execute the function invocation expression.
+            Value parentStrand = getParentStrand();
+            argValueList.add(0, parentStrand);
+            return argValueList;
+        } catch (ClassNotLoadedException e) {
+            throw new EvaluationException(String.format(EvaluationExceptionKind.FUNCTION_EXECUTION_ERROR.getString(),
+                    methodRef.name()));
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmInstanceMethod.java:105]: `getMethodArgs(JvmMethod method)` method is almost repeated in `JvmStaticMethod` and `JvmInstanceMethod`. Shall we add them to a Util class and use it from there?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmStaticMethod.java [hunk 6]
```diff
@@ -63,4 +71,40 @@ public Value invoke() throws EvaluationException {
...
+            // Todo - verify
+            // Here we use the parent strand instance to execute the function invocation expression.
+            Value parentStrand = getParentStrand();
+            argValueList.add(0, parentStrand);
+            return argValueList;
+        } catch (ClassNotLoadedException e) {
+            throw new EvaluationException(String.format(EvaluationExceptionKind.FUNCTION_EXECUTION_ERROR.getString(),
+                    methodRef.name()));
+        }
+    }
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmStaticMethod.java:109]: Repeated method mentioned in [here](https://github.com/ballerina-platform/ballerina-lang/pull/26322/files#r505451319)

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/EvaluationUtils.java (hunk 1 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/EvaluationUtils.java:56]: @Ibaqu These are being used outside of the class (inside the debugger evaluation engine implementation). `FROM_STRING_METHOD` and `FOR_NAME_METHOD` constants are used only inside the class and hence declared as private.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java (hunk 2 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java:120]: @praveennadarajah You might have missed that both of `DOUBLE_EQUAL_TOKEN` and `NOT_EQUAL_TOKEN` cases fall into `checkValueEquality` method.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java (hunk 3 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java:417]: This is intentional. This relevant method accepts two `java.lang.Object` instances as its arguments.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java (hunk 4 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java:439]: This is intentional. This relevant method accepts two java.lang.Object instances as its arguments.

Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java:439]: https://github.com/ballerina-platform/ballerina-lang/pull/26322#discussion_r505464641

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmInstanceMethod.java (hunk 5 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmInstanceMethod.java:105]: What happens inside the `getMethodArgs()` depends on the method type and therefore I've declared it as abstract in the parent class(`org.ballerinalang.debugadapter.evaluation.engine.JvmMethod`). WDYT?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmStaticMethod.java (hunk 6 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmStaticMethod.java:109]: https://github.com/ballerina-platform/ballerina-lang/pull/26322#discussion_r505469514

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java (hunk 2 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java:120]: Okay, Noted.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmInstanceMethod.java (hunk 5 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/JvmInstanceMethod.java:105]: +1 for declaring it as abstract in the parent class.

**Review by praveennadarajah — APPROVED** [review]

**Pull request merged**